### PR TITLE
Fix part of #59, part of #3926: Upgrade compute affected tests

### PR DESCRIPTION
**Pull request by BenHenning**
## Explanation
Fix part of #59
Fix part of #3926

As part of developing downstream PRs for #59, it was discovered that PRs which change a LOT of files (such as #4937) can run into problems where ComputeAffectedTests simply times out trying to compute the entire list of affected targets.

**Critical performance and compatibility fixes**

There have been past efforts to optimize the affected tests workflows (bucketing, breaking up some of the computations), but it was discovered that the most expensive part of the process is running the ``rbuildfiles`` query to figure out which BUILD files were affected by changed files. It was known this was an expensive step in the past, but it wasn't clear until this PR exactly how to address it. This PR changes the script to now:
- Filter ``rbuildfiles`` to only run under first-party targets (since it shouldn't be possible for third-party BUILD files to be affected by first-party changes). This reduces the search graph.
- Introduce Bazel command sharding for this step like the script already does for others. This breaks up the command to run on a subset of files multiple times, combining the output rather than running a single command with a large input. It seems that ``rbuildfiles`` does not scale linearly with its input size, so this drastically improves the script's performance. It's thought that this approach is also more logically correct due to more correct sibling matching semantics, but it's a bit hard to reason about ``bazel query`` behavior at times so I can't be 100% confident in this. Nevertheless, the existing tests pass and I haven't seen any issues from using these changes in downstream PRs.

Separately, another issue was discovered wherein some commands (including certain Bazel commands) can actually cause ``CommandExecutorImpl`` to soft-lock and always time out. This was due to an issue in the previous implementation wherein it would wait to read a command's stdout until after the timeout has been completed (i.e. it assumed the process would finish). This isn't correct, however: stdout is blocking I/O, and some commands are implemented to only continue execution after their standard output is read. The new implementation makes use of coroutine actors to consume stdout and stderr at the same time as waiting for execution to ensure all commands can continue execution and that they finish within the desired timeout. Note that the new ``ScriptBackgroundCoroutineDispatcher`` was actually introduced (in #5313) specifically to support this new ``CommandExecutorImpl`` implementation, though it has since been found to have lots of other nice benefits by providing scripts with a reliable mechanism for performing asynchronous operations without having to manage their own execution dispatcher.

Command execution for Bazel commands has also been updated to time out after 5 minutes rather than the previous 1 minute. Despite the optimizations and robustness improvements above, some commands still take quite some time to run for especially large and complex cases. While this change may result in a slower failure turnaround in cases when commands are soft-locked, it should result in better CI and script robustness in the long-term.

**Better support for chained PRs & possibly merge queues**

ComputeAffectedTests was also updated to use a merge base commit rather than a reference to the develop branch (where this new commit hash is provided by the CI workflow). The idea behind this is that the merge base commit is:
- More logically correct (as ``ComputeAffectedTests`` is meant to run in contexts where a branch wants to be merged into a destination).
- Better compatible with chained PRs. This allows for **significantly** better CI performance in chained PR situations since now only the tests relevant to the child PR will be run rather than all tests for the child & its parental hierarchy (see downstream PRs' CI runs to see this working in action).
- Hopefully better compatibility with merge queues (#3926). This hasn't been verified, but the flexibility in customizing the destination for affected tests should be the main prerequisite to properly supporting merge queues.

**Other changes**

``GitClient`` was updated to have a peace-of-mind check to ensure the base commit (provided as explained in the previous section) matches the merge base of the current branch. This should always be true (except maybe in merge queues--this will need to be verified). Note that this is only a soft warning, not an assertion failure.

``RepositoryFile`` was cleaned up slightly to be a bit more consistent with other directory management approaches done in scripts. I can see this being refactored more in the future. Callsite behavior isn't expected to be affected by these changes.

Some script tests were updated to have consistent formatting (which required updating the TODO exemptions). ``TodoIssueResolvedCheckTest`` and ``TodoOpenCheckTest`` also had some of their test file management cleaned up a bit.

**A note on testing**

These are inherently difficult things to test. I've verified what I could via CI and general observation, but I've also largely relied on existing tests to catch regressions (and many were caught during changes to the script). Since these are mainly implementation and not behavioral changes, I'm comfortable with the level of testing that was done.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- This is an infrastructure-only change.

### CODE REVIEW

**Comment by BenHenning** [comment]
Still active.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt [hunk 1]
```diff
@@ -45,9 +45,7 @@ class RetrieveLicenseTextsTest {
   private val outContent: ByteArrayOutputStream = ByteArrayOutputStream()
   private val originalOut: PrintStream = System.out
 
-  @Rule
-  @JvmField
-  var tempFolder = TemporaryFolder()
+  @field:[Rule JvmField] val tempFolder = TemporaryFolder()
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt:48]: TODO: document this & other clean up changes (also explains the todo exemption check updates).

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt [hunk 2]
```diff
@@ -23,9 +23,8 @@ class TodoIssueResolvedCheckTest {
     "Refer to https://github.com/oppia/oppia-android/wiki/Static-Analysis-Checks" +
       "#todo-issue-resolved-check for more details on how to fix this."
 
-  @Rule
-  @JvmField
-  var tempFolder = TemporaryFolder()
+  @field:[Rule JvmField]
+  val tempFolder = TemporaryFolder()
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt]: Move to single line for parity.

File: scripts/src/javatests/org/oppia/android/scripts/testing/TestBazelWorkspaceTest.kt [hunk 3]
```diff
@@ -98,6 +98,15 @@ class TestBazelWorkspaceTest {
     assertThat(workspaceFile.readLines()).isEmpty()
   }
 
+  @Test
+  fun testInitEmptyWorkspace_createsBazelVersionFileWithCorrectVersion() {
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/testing/TestBazelWorkspaceTest.kt]: TODO: remove this since it's now obsolete.

File: .github/workflows/unit_tests.yml [hunk 4]
```diff
@@ -76,10 +76,12 @@ jobs:
         id: compute-test-matrix
         env:
           compute_all_targets: ${{ contains(github.event.pull_request.title, '[RunAllTests]') }}
+          # See: https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request.
+          base_commit_hash: ${{ github.event.pull_request.base.sha }}
         # https://unix.stackexchange.com/a/338124 for reference on creating a JSON-friendly
         # comma-separated list of test targets for the matrix.
         run: |
-          bazel run //scripts:compute_affected_tests -- $(pwd) $(pwd)/affected_targets.log origin/develop compute_all_tests=$compute_all_targets
+          bazel run //scripts:compute_affected_tests -- $(pwd) $(pwd)/affected_targets.log $base_commit_hash compute_all_tests=$compute_all_targets
```
Comment (BenHenning) [inline .github/workflows/unit_tests.yml:84]: TODO: document this in PR description & its benefit for chained PRs.

TODO: verify whether this is sufficient for merge queues.

File: scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt [hunk 5]
```diff
@@ -33,17 +43,57 @@ class CommandExecutorImpl(
...
-    val finished = runBlocking {
-      CoroutineScope(scriptBgDispatcher).async {
-        process.waitFor(processTimeout, processTimeoutUnit)
-      }.await()
+
+    // Consume the input & error streams individually, and separately from waiting for the process
+    // to complete (since consuming the output channels may be required for the process to actually
+    // finish executing properly).
+    val stdoutLinesDeferred = process.inputStream.readAllLinesAsync()
+    val stderrLinesDeferred = process.errorStream.readAllLinesAsync()
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt:51]: TODO: explain these executor changes (& why they're the reason why the script background executor was added in the first place).

File: scripts/src/java/org/oppia/android/scripts/common/GitClient.kt [hunk 6]
```diff
@@ -36,7 +36,11 @@ class GitClient(
   }
 
   private fun retrieveBranchMergeBase(): String {
-    return executeGitCommandWithOneLineOutput("merge-base $baseDevelopBranchReference HEAD")
+    return executeGitCommandWithOneLineOutput("merge-base $baseCommit HEAD").also {
+      if (baseCommit != it) {
+        println("WARNING: Provided base commit $baseCommit doesn't match merge-base: $it.")
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/GitClient.kt:41]: TODO: explain in PR description why this change was added (peace of mind due to functional differences from previous version that checks against the develop branch).

File: scripts/src/java/org/oppia/android/scripts/common/RepositoryFile.kt [hunk 7]
```diff
@@ -42,33 +42,25 @@ class RepositoryFile() {
...
       // Note that Files.walk() is used instead of Kotlin's walk() function since the latter follows
       // symbolic links which is almost 10x slower than not following them (due to very deep Bazel
       // build directories), and it's not necessary to follow the symlinks.
-      return Files.walk(File(repoPath).toPath()).asSequence().map(Path::toFile).filter { file ->
-        val isProhibited = checkIfProhibitedFile(retrieveRelativeFilePath(file, repoPath))
-        !isProhibited &&
+      val repoRoot = File(repoPath).absoluteFile.normalize()
+      return Files.walk(repoRoot.toPath()).asSequence().map(Path::toFile).map { file ->
+        file.absoluteFile.normalize()
+      }.filter { file ->
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/RepositoryFile.kt:53]: TODO: mention that there are some basic cleanup efforts here to simplify the exemption checking a bit. Shouldn't change callsite behavior.

File: scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt [hunk 8]
```diff
@@ -57,16 +54,19 @@ class BazelClient(
     // Note that this check is needed since rbuildfiles() doesn't like taking an empty list.
     return if (buildFileList.isNotEmpty()) {
       val referencingBuildFiles =
-        executeBazelCommand(
-          "query",
+        runPotentiallyShardedQueryCommand(
+          "filter('^[^@]', rbuildfiles(%s))", // Use a filter to limit the search space.
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt:58]: TODO: Mention this in PR description. It's a significant reason why ComputeAffectedTests now works correctly for very large PR results since this was found to be the primary bottleneck in performance. Sharding should be more or less the same in behavior (see comment below).

File: scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt [hunk 9]
```diff
@@ -28,27 +29,28 @@ private const val MAX_TEST_COUNT_PER_SMALL_SHARD = 15
...
  *   bazel run //scripts:compute_affected_tests -- $(pwd) /tmp/affected_test_buckets.proto64 \\
- *     origin/develop compute_all_tests=false
+ *     abcdef0123456789 compute_all_tests=false
  */
 fun main(args: Array<String>) {
   if (args.size < 4) {
     println(
       "Usage: bazel run //scripts:compute_affected_tests --" +
-        " <path_to_directory_root> <path_to_output_file> <base_develop_branch_reference>" +
+        " <path_to_directory_root> <path_to_output_file> <merge_base_commit>" +
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt:45]: TODO: Document this change (that the merge base branch is now provided by the GitHub workflow).

File: scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt [hunk 10]
```diff
@@ -81,7 +83,10 @@ class ComputeAffectedTests(
   val maxTestCountPerLargeShard: Int = MAX_TEST_COUNT_PER_LARGE_SHARD,
   val maxTestCountPerMediumShard: Int = MAX_TEST_COUNT_PER_MEDIUM_SHARD,
   val maxTestCountPerSmallShard: Int = MAX_TEST_COUNT_PER_SMALL_SHARD,
-  val commandExecutor: CommandExecutor = CommandExecutorImpl(scriptBgDispatcher)
+  val commandExecutor: CommandExecutor =
+    CommandExecutorImpl(
+      scriptBgDispatcher, processTimeout = 5, processTimeoutUnit = TimeUnit.MINUTES
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt:88]: TODO: document this behavior change.

**Review by BenHenning — COMMENTED** [review]
Self-review changes.

**Review by BenHenning — COMMENTED** [review]
Reviewed post-merge changes.

File: scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt (hunk 2 above)
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/todo/TodoIssueResolvedCheckTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/testing/TestBazelWorkspaceTest.kt (hunk 3 above)
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/testing/TestBazelWorkspaceTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt (hunk 8 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/BazelClient.kt:58]: Done.

File: scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt (hunk 5 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/CommandExecutorImpl.kt:51]: Done.

File: scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt (hunk 10 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt:88]: Done.

File: scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt (hunk 9 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/ci/ComputeAffectedTests.kt:45]: Done.

File: .github/workflows/unit_tests.yml (hunk 4 above)
Comment (BenHenning) [reply .github/workflows/unit_tests.yml:84]: Skipping the merge queues bit since that's part of #3926. Description otherwise updated.

File: scripts/src/java/org/oppia/android/scripts/common/GitClient.kt (hunk 6 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/GitClient.kt:41]: Done.

File: scripts/src/java/org/oppia/android/scripts/common/RepositoryFile.kt (hunk 7 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/common/RepositoryFile.kt:53]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt (hunk 1 above)
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt:48]: Done.

**Review by BenHenning — COMMENTED** [review]
Self-reviewed latest changes.

**Comment by BenHenning** [comment]
@adhiamboperes PTAL.

**Review by adhiamboperes — APPROVED** [review]
Thanks @BenHenning! These chnages LGTM.

**Comment by adhiamboperes** [comment]
Please feel free to merge this PR.

**Comment by BenHenning** [comment]
Thanks @adhiamboperes! Will be merging this shortly.
